File: real_benchmarks/data/real_c_rust_pairs/union_type_punning.c
/*
 * Union type punning divergence — inspired by real networking/serialization code.
 *
 * In C, type-punning through unions is explicitly legal (C11 §6.5.2.3/3, footnote 95).
 * In Rust, unions require unsafe and reading wrong variant is UB per Rust spec.
 * This is a fundamental semantic divergence: legal C → UB Rust.
 */
#include <stdint.h>
#include <string.h>

/* Type punning: reinterpret float bits as integer (legal in C). */
union FloatBits {
    float f;
    uint32_t u;
};

uint32_t float_to_bits(float value) {
    union FloatBits fb;
    fb.f = value;
    return fb.u;  /* Legal in C11 — reads different member than last written */
}

float bits_to_float(uint32_t bits) {
    union FloatBits fb;
    fb.u = bits;
    return fb.f;  /* Legal type pun in C */
}

/* Network byte-order conversion using union (common real-world pattern). */
union NetWord {
    uint32_t word;
    uint8_t bytes[4];
};

uint32_t ntohl_manual(uint32_t net_order) {
    union NetWord nw;
    nw.word = net_order;
    /* Read individual bytes — legal union access in C */
    return ((uint32_t)nw.bytes[0] << 24) |
           ((uint32_t)nw.bytes[1] << 16) |
           ((uint32_t)nw.bytes[2] << 8)  |
           ((uint32_t)nw.bytes[3]);
}

/* Tagged union pattern (discriminated union). */
struct Variant {
    int tag;  /* 0 = int, 1 = float */
    union {
        int32_t i;
        float f;
    } data;
};

int32_t extract_int(struct Variant *v) {
    if (v->tag == 0)
        return v->data.i;
    /* Reading .i when .f was last written — still legal in C */
    return v->data.i;
}

File: real_benchmarks/data/real_c_rust_pairs/union_type_punning.rs
/*
 * Rust translation of union type punning patterns.
 * In Rust, union field access requires unsafe.
 * The idiomatic approach is transmute or to_bits()/from_bits().
 */

/// Float to bits: Rust has f32::to_bits() — safe, no union needed.
pub fn float_to_bits(value: f32) -> u32 {
    value.to_bits()  // safe, returns IEEE-754 representation
}

/// Bits to float: Rust has f32::from_bits() — safe.
pub fn bits_to_float(bits: u32) -> f32 {
    f32::from_bits(bits)  // safe, defined behavior
}

/// Network byte order: Rust uses u32::from_be_bytes / to_be_bytes.
/// No union type punning needed.
pub fn ntohl_manual(net_order: u32) -> u32 {
    let bytes = net_order.to_ne_bytes();
    ((bytes[0] as u32) << 24)
        | ((bytes[1] as u32) << 16)
        | ((bytes[2] as u32) << 8)
        | (bytes[3] as u32)
}

/// Tagged union: Rust enum is safe — discriminant is checked by compiler.
pub enum Variant {
    Int(i32),
    Float(f32),
}

pub fn extract_int(v: &Variant) -> i32 {
    match v {
        Variant::Int(i) => *i,
        Variant::Float(f) => *f as i32,  // explicit conversion, not type punning
    }
}
